payu/branch.py

```diff
@@ -0,0 +1,375 @@
+"""Experiment branch support for payu's branch, clone and checkout commands
+
+This may generate new experiment ID, updates, sets any
+specified configuration in config.yaml and updates work/archive symlinks
+
+:copyright: Copyright 2011 Marshall Ward, see AUTHORS for details.
+:license: Apache License, Version 2.0, see LICENSE for details.
+"""
+
+import os
+import warnings
+from pathlib import Path
+import re
+from typing import Optional
+
+from ruamel.yaml import YAML, CommentedMap
+import git
+
+from payu.fsops import read_config, DEFAULT_CONFIG_FNAME
+from payu.laboratory import Laboratory
+from payu.metadata import Metadata, UUID_FIELD, METADATA_FILENAME
+from payu.git_utils import GitRepository, git_clone
+
+
+NO_CONFIG_FOUND_MESSAGE = """No configuration file found on this branch.
+Skipping adding new metadata file and creating archive/work symlinks.
+
+To try find a branch that has config file, you can:
+    - Display local branches by running:
+        payu branch
+    - Or display remote branches by running:
+        payu branch --remote
+
+To checkout an existing branch, run:
+    payu checkout BRANCH_NAME
+Where BRANCH_NAME is the name of the branch"""
+
+
+def archive_contains_restarts(archive_path: Path) -> bool:
+    """Return True if there's pre-existing restarts in archive"""
+    pattern = re.compile(r"^restart[0-9][0-9][0-9]+$")
+    if not archive_path.exists():
+        return False
+
+    for path in archive_path.iterdir():
+        real_path = path.resolve()
+        if real_path.is_dir() and pattern.match(path.name):
+            return True
+    return False
+
+
+def check_restart(restart_path: Optional[Path],
+                  archive_path: Path) -> Optional[Path]:
+    """Checks for valid prior restart path. Returns resolved restart path
+    if valid, otherwise returns None"""
+    if restart_path is None:
+        return
+
+    # Check for valid path
+    if not restart_path.exists():
+        warnings.warn((f"Given restart path {restart_path} does not "
+                       f"exist. Skipping setting 'restart' in config file"))
+        return
+
+    # Resolve to absolute path
+    restart_path = restart_path.resolve()
+
+    # Check for pre-existing restarts in archive
+    if archive_contains_restarts(archive_path):
```

Comment:
If `archive_path` does not exist at this point, adding the  `restart: restart_path`  to `config.yaml` still works. This check for pre-existing restarts was to prevent adding `restart: restart_path` flag, as an existing restart would be used instead.

For new experiments the `archive_path` is always created at a later point - when the metadata is copied over (after it has been created)

Reply:
Cool. I clearly didn't (don't!) understand, sorry if this was a waste of time.